src/main/java/com/linkedin/kmf/services/ProduceService.java

```diff
@@ -154,15 +156,44 @@ public void start() {
   }
 
   private void initializeStateForPartitions() {
+    Map<Integer, String> keyMapping = generateKeyMappings();
     int partitionNum = _partitionNum.get();
     for (int partition = 0; partition < partitionNum; partition++) {
+      String key = keyMapping.get(partition);
       //This is what preserves sequence numbers across restarts
-      if (!_nextIndexPerPartition.containsKey(partition)) {
-        _nextIndexPerPartition.put(partition, new AtomicLong(0));
+      if (!_nextIndexPerKey.containsKey(key)) {
+        _nextIndexPerKey.put(key, new AtomicLong(0));
         _sensors.addPartitionSensors(partition);
       }
-      _produceExecutor.scheduleWithFixedDelay(new ProduceRunnable(partition), _produceDelayMs, _produceDelayMs, TimeUnit.MILLISECONDS);
+      _produceExecutor.scheduleWithFixedDelay(new ProduceRunnable(partition, key), _produceDelayMs, _produceDelayMs, TimeUnit.MILLISECONDS);
+    }
+  }
+
+  private Map<Integer, String> generateKeyMappings() {
+    int partitionNum = _partitionNum.get();
+    HashMap<Integer, String> keyMapping = new HashMap<>();
+
+    int keyIndex = 0;
+    while (keyMapping.size() < partitionNum) {
+      String key = keyIndex + "";
+      int partition = getDefaultPartitionForKey(key);
+      if (!keyMapping.containsKey(partition)) {
+        keyMapping.put(partition, key);
+      }
+      keyIndex++;
     }
+
+    return keyMapping;
+  }
+
+  //TODO: Remove method and convert uses to Util.toPositive when kafka version dependency reaches 0.10.1.10
+  public static int toPositive(int number) {
+    return number & 0x7fffffff;
+  }
+
+  private int getDefaultPartitionForKey(String key) {
```

Comment:
We probably need a way for user to configure the partitioning algorithm. The patch currently assumes that MM will only use the new producer's partitioning algorithm. This is OK for users of vanilla Kafka. But we actually have hotfix in LinkedIn which changes the MM to use old producer's partitioning algorithm. This is because the algorithm used by old producer and new producer is different and we may cause message re-order if we simply change the algorithm.

We can also user to configure this similar to `produce.producer.class` in `ProduceServiceConfig`.